lib/processor/http.go

```diff
@@ -205,7 +214,7 @@ func (h *HTTP) ProcessMessage(msg types.Message) ([]types.Message, types.Respons
 			}
 			h.mErr.Incr(1)
 			h.mErrHTTP.Incr(1)
-			h.log.Errorf("HTTP request to '%v' failed: %v\n", h.conf.HTTP.URL, err)
+			h.log.Errorf("HTTP request to '%v' failed: %v\n", h.urlField.String(0, msg), err)
```

Comment:
Unfortunately we can't assume that the interpolation here will be the exact same as the actual URL, since there could be an aspect of randomness to it. I think the best option here is to instead add the URL as a prefix to the error string returned from `h.client.Send`, so here we would just print `HTTP request failed: %v\n` and the returned error would be some thing like `http://example.com/foobar: big foot is real`.

Reply:
Oh, interesting! But I'm not sure I get it. For example for the 404 case, the error returned by `h.client.Send` doesn't contain the actual URL, just "HTTP request returned unexpected response code (404): 404 Not Found\n". I can add a defer inside `h.client.DoWithContext` which pulls the URL directly from the `req` object (if it's not nil) and prepends it to the error, but I'm not sure if that's what you're suggesting.

I looked through the code and the URL is constructed [here](https://github.com/Jeffail/benthos/blob/58453c21ad1bd09ab94532e4ab85f1a454f1ec36/lib/util/http/client/type.go#L356) where the `url` `field.Expression` is stringified using the current message, but since `CreateRequest` is only called within `h.client.DoWithContext`, I don't see what would cause it to add randomness to the URL. Does it have something to do with the retry logic? I'm thinking to add a test for this.

Reply:
The randomness is because you can use functions like `now()` and `random_int()` that'll resolve differently each time the interpolation is made. I think the approach I'd lean towards is adding:

```go
	var req *http.Request
	if req, err = h.CreateRequest(msg); err != nil {
		h.mErrReq.Incr(1)
		h.mErr.Incr(1)
		logErr(err)
		return nil, err
	}
	defer func() {
		if err != nil {
			err = fmt.Errorf("%v: %w", req.URL, err)
		}
	}()
```

There are some `return nil, types.ErrTypeClosed` cases in there but those are probably fine to return without wrapping as it's unrelated to the URL. There's a line `return nil, err` near the bottom that should probably be changed for just `return` as I think that would also bypass the defer.

Reply:
Oh, that makes sense, thank you! Forgot that bloblang can do everything 😅 Changes pushed.